### DISCUSSION TITLE
Ainda vale a pena utilizar leiningen para projetos novos?

### DISCUSSION THREAD

**Discussion by wandersoncferreira** [clj-br/forum#4] [Geral]
Não sei dizer se a adoção do `deps.edn` para gerenciar projetos tem sido bem vista pela comunidade como um todo, mas me parece bem mais simples para iniciantes. Quais são suas opiniões sobre lein vs deps?

**Comment by ericdallo** [comment]
Prefiro `deps.edn` com alguns pontos:

- o CLI do `clojure` não é tao simples para quem ta começando quanto o `lein`, mas a longo prazo acho melhor o `deps.edn`
- o `deps.edn` não tem um test runner tão bom quanto lein, embora existam libs como kaocha que faz funcionar bem.
- o `deps.edn` por enquanto não pode gerar jars built-in, precisando de algo como `depstar` por exemplo. Logo o `tools.build` deve resolver isso.

Apesar destes pontos, eu prefiro bem mais o deps.edn por ter uma API/configuração mais simples/extensível :)

**Reply by pbalduino** [reply]
Como fica o caso em que eu preciso compilar uma ou mais classes Java dentro do projeto Clojure?

O `Lein` trata isso de forma bem simples, mas tive problemas quando tentei fazer com `deps.edn`.

**Reply by wandersoncferreira** [reply]
Você pode apontar um exemplo de como o `lein` faz isso?

**Reply by ericdallo** [reply]
Eu tb tive este exato problema @pbalduino, por algum tempo extrai as classes para uma jar/projeto separado até criarem o [americano](https://github.com/IGJoshua/americano) que funciona perfeitamente (e até mais rapido que o lein) :)
Um exemplo em que usei essa lib: https://github.com/clojure-lsp/clojure-lsp/blob/master/deps.edn#L23-L25

Alem do mais, Isso provavelmente ira vir no futuro `tools.build`

**Reply by ericdallo** [reply]
@wandersoncferreira o `lein` faz isso somente add `:src-java` em seu `project.clj` até onde lembro

**Comment by lsevero** [comment]
Eu continuo preferindo o lein para gerenciar os projetos. O lein tem uma interface centralizada e única para várias funcionalidades, e por centralizar as funcionalidades também centraliza a documentação delas. No deps precisamos de plugins para basicamente tudo, test runners, deploy, geração de uberjars, compilação de classes java, checagem de compilação e por ai vai, para cada plugin é um repositorio diferente para ler uma documentação. Tudo isso vem de fábrica no lein, e todos os projetos do lein usam o mesmo CLI. É muito comum os projetos do deps terem bash scripts ou makefiles para listar todos os comandos usaveis, pois só de ler um deps.edn não da pra saber de cara qual comando faz o quê.

A única vantagem que vejo no deps ainda é o gerenciamento de dependencias, em casos muito específicos pode ser útil usar dependências locais ou de um git, ainda assim isso deve ser evitado ao máximo e é sempre preferível puxar dependências do mvn repository ou do clojars.

O lein consegue ser feature complete sem ser inchado, a maioria dos projetos usando o deps acaba adicionando plugins até ter quase todas as features que o lein já oferece, e acaba perdendo padronização.

**Comment by mauricioszabo** [comment]
Eu confesso que não entendi essa frase:

> me parece bem mais simples para iniciantes

O que pode ser mais fácil que `lein new app <nome-da-app>`?

**Reply by wandersoncferreira** [reply]
> O que pode ser mais fácil que `lein new app <nome-da-app>`?

Não precisar entender o resultado desse comando. Criar uma pasta vazia e um arquivo `deps.edn` com

```clojure
{:paths ["src"]}
```

parece ser o mais simples que dá pra ficar.

**Reply by mauricioszabo** [reply]
Continuo não entendendo. "Criar um arquivo vazio e adicionar um texto" não é mais fácil que ter um gerador que gera o arquivo pra ti. Em `lein` vc vai ter que entender o formato do arquivo `project.clj`. Em deps, o formato do arquivo `deps.edn`. Realmente, não entendi o ponto.

De qq maneira, vc precisa entender o que é `:paths`. E o formato das deps. E se vc vai adicionar uma main function, precisa também saber como funciona. O `lein` já te dá isso de graça - `lein run` já roda o app pra ti, `lein test` já roda os testes, e vc já tem o `src` criado e tal.
